xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -373,33 +431,65 @@ private Set<String> getResourceKeys(XdsResourceType<?> xdsResourceType) {
     return resourceSubscribers.get(xdsResourceType).keySet();
   }
 
-  private void cleanUpResourceTimers() {
+  // cpcForThisStream is null when doing shutdown
+  private void cleanUpResourceTimers(ControlPlaneClient cpcForThisStream) {
+    Collection<String> authoritiesForCpc = getActiveAuthorities(cpcForThisStream);
+
     for (Map<String, ResourceSubscriber<?>> subscriberMap : resourceSubscribers.values()) {
       for (ResourceSubscriber<?> subscriber : subscriberMap.values()) {
-        subscriber.stopTimer();
+        if (cpcForThisStream == null || authoritiesForCpc.contains(subscriber.authority)) {
+          subscriber.stopTimer();
+        }
       }
     }
   }
 
-  public ControlPlaneClient getOrCreateControlPlaneClient(ServerInfo serverInfo) {
+  private ControlPlaneClient getOrCreateControlPlaneClient(String authority) throws IOException {
+    // Optimize for the common case of a working ads stream already exists for the authority
+    ControlPlaneClient activeCpc = getActiveCpc(authority);
+    if (activeCpc != null && !activeCpc.isInBackoff()) {
+      return activeCpc;
+    }
+
+    ImmutableList<ServerInfo> serverInfos = getServerInfos(authority);
+    if (serverInfos == null) {
+      throw new IllegalArgumentException("No xds servers found for authority " + authority);
+    }
+
+    for (ServerInfo serverInfo : serverInfos) {
```

Comment:
This loop can consider cpcs that aren't currently registered for the authority, but it won't add them to `activatedCpClients`. That can cause `activatedCpClients` to be corrupted, where cpcs should be registered for the authority but aren't.

Reply:
It is the manageControlPlaneClient method that is responsible for calling addCpcToAuthority which will add the specified cpc and any missing higher priority ones.  So, by the end of manageControlPlaneClient, activatedCpClients will have exactly the cpcs that it should.